docs/blogs/ata_elastic_scheduling.md

```diff
@@ -0,0 +1,129 @@
+# ElasticDL：同时提升深度学习作业的研发效率和集群利用率
+
+## 背景
+
+Kubernetes 是目前最先进的分布式操作系统，是公有云和私有云的事实工业标准。
+目前，在 Kubernetes 机群上运行 ***分布式*** TensorFlow 作业主要依赖 Kubeflow 项目
+提供的 Kubernetes controller —— [kubeflow/tf-operator](https://github.com/kubeflow/tf-operator)
+—— 这是因为 TensorFlow 不是一个 Kubernetes-native 的编程框架，自己不会调用
+Kubernetes API 启动进程。类似的，很多诞生在 Kubernetes 之前的分布式编程框架
+（比如 Spark）都需要一个 Kubernetes controller 来启动作业。这类应用或者框架
+特定的 controller 被 CoreOS 公司的工程师们称为 Kubernetes Operator，后来
+演化成了行业术语。
+
+和早年间用 MPI 运行分布式深度学习作业一样，kubeflow/tf-operator 并不支持
+弹性调度，而是启动 N 个进程去执行作业。如果执行中途有一些进程被
+抢占（preempted）或者失败（fail），则作业失败，只能从最近的 checkpoint 重启。
+这样的调度模式被称为 [gang scheduling](https://en.wikipedia.org/wiki/Gang_scheduling)。
```

Comment:
Kubeflow/tf-operator 也不直接支持 gang scheduling的，还需要安装一个额外的volcano scheduler。请参考 volcano scheduler

Reply:
那不安装 volcano scheduler 可以执行 TF job 吗？

Reply:
这是安装volcano scheduler 之后的效果
> With using volcano scheduler to apply gang-scheduling, a job can run only if there are enough resources for all the pods of the job. Otherwise, all the pods will be in pending state waiting for enough resources. For example, if a job requiring N pods is created and there are only enough resources to schedule N-2 pods, then N pods of the job will stay pending.

我推断，如果不安装 volcano scheduler的话，一个job请求N个 pod，但是剩余资源只够 N-2 个pod，Kubeflow如果继续执行 TF job会失败。

Reply:
应该是使用 volcano 的话，如果需要 N 个 pod，要么 N 个 pod 一起启动，要么 N 个 pod 全部pending。如果不适用 volcano，kubeflow 会尝试启动 N 个 pod，如果只够起 N - 2 个 pod，则就先起 N - 2 个 pod，但是这个 N - 2 个 pod 里启动的 TensorFlow 程序并不会开始模型迭代，而是等待剩下两个 pod 启动后才开始。但是先启动的 N -2 个 pod 已经被占用了，不能被其他任务使用，会造成资源浪费。使用 volcano 不存在这种资源浪费。